lib/govuk_index/presenters/method_builder.rb

```diff
@@ -0,0 +1,26 @@
+module GovukIndex
+  module MethodBuilder
+    def add_method(name, field_name = name)
+      define_method name do
+        payload[field_name.to_s]
+      end
+
+      eval <<-METHOD
+        def #{name}
+          #{data_object}[#{field_name}]
+        end
+      METHOD
+    end
+
+    def data_object
+      @data_object || :payload
+    end
+
+    def data_object=(data_object)
+      define_method :payload do
+        public_send(data_object)
+      end
+      # @data_object = data_object
+    end
+  end
```

Comment:
this is a combination of approaches as we are trying to decide on an approach

Reply:
As someone who doesn't understand a lot of ruby metaprogramming, I'd go with what looks simplest: `define_method`.

Another approach I've seen is to call attr_reader with an array of names. If you have a list of all things you want to delegate to the hash, then you can do something like this

```ruby
attr_reader *FIELDS
```

or 

```ruby
attr_accessor *FIELDS
```

Then you can do all your metaprogramming in the initializer:

```ruby
FIELDS.each do |field|
  instance_variable_set field payload[field]
end
```

I don't have a strong opinion on this, but we should keep it as simple as possible, and make sure that the syntactic sugar doesn't get in the way of understanding how the presenters are mapping from one to thing to the other.

:man_shrugging: